app/uk/gov/hmrc/gform/gformbackend/GformConnector.scala

```diff
@@ -137,9 +137,9 @@ class GformConnector(ws: WSHttp, baseUrl: String) {
   //TODO move this file to gform and make it's origin there
 
   /****** Tax Period ******/
-  def getAllTaxPeriods(htps: NonEmptyList[HmrcTaxPeriod])(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
+  def getAllTaxPeriods(htps: List[HmrcTaxPeriodWithEvaluatedId])(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
```

Comment:
What was the need to revert back to `List`?

Reply:
Using NonEmptyLists was causing issues as in several instances we were needing to filter over the NonEmptyLists which would then give a List which we would then need to convert back to a NonEmptyList etc.

Reply:
Whenever one needs to filter on NonEmptyList result cannot be safely turned back into NonEmptyList. Desire so is indication that something is probably wrong with the design. 

The solution to get rid of NonEmptyList on input is least preferred one. Cannot we just keep working with List after the filter?

Reply:
The issue is that the NonEmptyList is filtered before this point, making it a list. As such this section of the code would need to take a List rather than a NonEmptyList even if we used NonEmptyList earlier

Reply:
You should not be calling the backend via uk.gov.hmrc.gform.gformbackend.GformConnector#getAllTaxPeriods (causing an inter-microservice http call) if the list of HmrcTaxPeriodWithEvaluatedId is empty, it will have no effect and you will simply be getting back an empty list of TaxResponse at considerable cost

Reply:
Despite not taking a NonEmptyList we still should never call the backend with an empty list as the list we pass to getAllTaxPeriods is formed from a NonEmptyList. The reason it becomes a list is that trying to traverse over the NonEmptyList to evaluate the IdNumbers doesn't seem to work with a NonEmptyList, so I have used .toList on it

Reply:
import cats.implicits._

...

hmrcTaxPeriodIdentifiers.nonEmptyTraverse(